classes/04class/exercises/c04-iac02/danielvca/vpc.tf

```diff
@@ -0,0 +1,127 @@
+resource "aws_vpc" "devopsacademyiac" {
+  cidr_block = var.vpc_cidr
+
+  tags = {
+    Name = "devopsacademy-iac"
+  }
+}
+
+resource "aws_network_acl" "main" {
+  vpc_id = aws_vpc.devopsacademyiac.id
+
+  egress {
+    protocol   = "all"
+    rule_no    = 100
+    action     = "allow"
+    cidr_block = "10.0.0.0/16"
+    from_port  = 0
+    to_port    = 0
+  }
+
+  egress {
+    protocol   = "all"
+    rule_no    = 200
+    action     = "allow"
+    cidr_block = "180.150.58.136/32"
+    from_port  = 0
+    to_port    = 0
+  }
+
+  ingress {
+    protocol   = "all"
+    rule_no    = 300
+    action     = "allow"
+    cidr_block = "10.0.0.0/16"
+    from_port  = 0
+    to_port    = 0
+  }
+
+  ingress {
+    protocol   = "all"
+    rule_no    = 400
+    action     = "allow"
+    cidr_block = "180.150.58.136/32"
+    from_port  = 0
+    to_port    = 0
+  }
+
+}
+
+resource "aws_internet_gateway" "internet_gw" {
+  vpc_id = aws_vpc.devopsacademyiac.id
+}
+
+resource "aws_subnet" "private_subnet_a" {
+  vpc_id            = aws_vpc.devopsacademyiac.id
+  cidr_block        = var.subnets.private_subnet_a.cidr
+  availability_zone = var.subnets.private_subnet_a.az
+}
```

Comment:
You can practise **DRY** (**D**on't **R**epeat **Y**ourself 😆) and use `for_each` here like:

```hcl
resource "aws_subnet" "private_subnet" {
  for_each = var.subnets.private_subnet

  vpc_id            = aws_vpc.devopsacademyiac.id
  cidr_block        = each.value.cidr
  availability_zone = each.value.az
  tags = {
     Name = each.key # or each.value.name
  }
}
```

you would need to split your subnets in two variables:

```hcl
private_subnets = {
  "private_subnet_a" : {
    name = "private_subnet_a",
    cidr = "10.0.0.0/24",
    az   = "ap-southeast-2a"
  },
  "private_subnet_b" : {
    name = "private_subnet_b",
    cidr = "10.0.1.0/24",
    az   = "ap-southeast-2b"
  }
}

public_subnets = {
  "public_subnet_a" : {
    name = "public_subnet_a",
    cidr = "10.0.2.0/24",
    az   = "ap-southeast-2a"
  },

  "public_subnet_b" : {
    name = "public_subnet_b",
    cidr = "10.0.3.0/24",
    az   = "ap-southeast-2b"
  }
}
```

The same concept would be applied for your `aws_route_table_assossiation`, you would need only one.

More info to how to use for_each here: https://www.terraform.io/docs/configuration/resources.html#for_each-multiple-resource-instances-defined-by-a-map-or-set-of-strings

> You could go even further and use [`cidrsubnet`](https://www.terraform.io/docs/configuration/functions/cidrsubnet.html) to calculate your subnet cidr_block based on the vpc cidr